jmclient/jmclient/wallet_service.py

```diff
@@ -347,23 +360,31 @@ def transaction_monitor(self):
                 # is absurdly fast, this is considered acceptable compared with
                 # additional complexity.
                 self.log_new_tx(removed_utxos, added_utxos, txid)
-                self.processed_txids.append(txid)
+                self.processed_txids.add(txid)
 
             # first fire 'all' type callbacks, irrespective of if the
             # transaction pertains to anything known (but must
             # have correct label per above); filter on this Joinmarket wallet label,
             # or the external monitoring label:
-            if (self.bci.is_address_labeled(tx, self.get_wallet_name()) or
-                self.bci.is_address_labeled(tx, self.EXTERNAL_WALLET_LABEL)):
+            if (any(self.bci.is_address_labeled(addr, wallet_name) or
+                    self.bci.is_address_labeled(addr,
+                        self.EXTERNAL_WALLET_LABEL)
+                    for addr in res["details"])):
                 for f in self.callbacks["all"]:
                     # note we need no return value as we will never
                     # remove these from the list
                     f(txd, txid)
 
-            # The tuple given as the second possible key for the dict
-            # is such because txid is not always available
-            # at the time of callback registration).
-            possible_keys = [txid, tuple((x.scriptPubKey, x.nValue) for x in txd.vout)]
+            # txid is not always available at the time of callback registration.
+            # Migrate any callbacks registered under the provisional key, and
+            # leave a txid breadcrumb so check_callback_called can find it.
```

Comment:
Just to be clear: makers were the ones previously using the `txinfo` tuple instead of the `txid` as key for their callback lists (see the comment on line 499 of client_protocol), but - they still kept track of the callbacks alright? That is, `check_callback_called` was scheduled after `unconfirm_timeout_seconds` and then after `confirm_timeout_hours`, in both cases using the `txinfo` tuple right?
I think your 'breadcrumb' approach makes sense, but I want to be clear whether you're saying the existing tracking was not working? I *think* the bugs you referred to in the main commit comment are separate to this? But I'm not sure.

Reply:
I didn't look too carefully at who is using the callback mechanisms, but I did see that makers schedule a check for some time after registering the callback to check that the callback did fire. (Actually, the default timeout of 3 minutes seems a little too short in my real-world usage, as takers don't always get around to broadcasting the transaction within 3 minutes. I bumped my timeout setting up to 5 minutes.) And yes, the callbacks are initially registered using the provisional key, not the TxID. As soon as the transaction is observed in the Bitcoin Core wallet, the callback registrations are migrated over to the TxID, and the breadcrumb is left at the provisional key to support deregistering under that key.

Something I don't understand is why the callbacks don't always register using the TxID. Maybe it's an artifact left over from the pre-SegWit days when signing the transaction would change its TxID? That's not the case anymore, though, so the TxID is now known as soon as the unsigned transaction is fully constructed and does not change as signatures are added.

> I want to be clear whether you're saying the existing tracking was not working?

Existing tracking was working, though it was only bug-free for a single callback of each type per transaction. I think that's all that currently gets registered, so the logic bugs weren't manifesting, but if multiple callbacks of a given type were ever registered for a given key, or if a `confirmed` callback was registered without an `unconfirmed` callback also being registered for the same key, then it would not have worked correctly.

Reply:
> Something I don't understand is why the callbacks don't always register using the TxID. Maybe it's an artifact left over from the pre-SegWit days when signing the transaction would change its TxID? That's not the case anymore, though, so the TxID is now known as soon as the unsigned transaction is fully constructed and does not change as signatures are added.

Yes you're right. It was quite a long while ago but when refactoring with this WalletService object, I didn't pay attention to the fact that even for wrapped segwit, it was no longer the case that we needed this logic. At that point non-segwit may have still existed at just a vestigial level, not sure. I suppose it's strictly better to leave it in case we ever have a need to use non-segwit transactions, though it would be hard to argue against a removal of it, either.

Reply:
> Existing tracking was working, though it was only bug-free for a single callback of each type per transaction. I think that's all that currently gets registered, so the logic bugs weren't manifesting, but if multiple callbacks of a given type were ever registered for a given key, or if a confirmed callback was registered without an unconfirmed callback also being registered for the same key, then it would not have worked correctly.

Thanks, I see you have addressed this question in a separate comment.

Reply:
@whitslack 

I was troubled this morning that I'd forgotten something, and in fact I had: for p2sh-wrapped segwit, the scriptSig is *not* empty in the complete tx - it contains a push of the witness program which is the redeem script of the p2sh output being spent. This means that the `txid` field for the transaction (see BIP141 for this; and note, `txid`, not `wtxid`), which hashes in the list of txins, is affected by the redeemscripts of the outputs being spent. So the unsigned transaction's hash with scriptSigs blanked out, that the maker sees (and note: it is fine for signing, because the *signature hash* does not include the contents of the scriptSig) is not the same as the final `txid`. This doesn't affect native segwit of course, because the scriptSig is still blanked out even in the final version.

I think basically nobody is using p2sh-wrapped segwit any more, but it is still supported.